File: src/adapters/mal_client.rs
use loco_rs::prelude::async_trait;
use loco_rs::{Error, Result};
use serde_json::{Value, json};
use tracing::debug;

use crate::{
    adapters::{
        MediaClient, MediaClientParams,
        mal_models::{MalResponse, MalStatus},
    },
    core::{ResultExt, constants::DEFAULT_HOSTED_AUTH_PAGE},
    models::{
        media::{MediaProvider, PaginatedResponse},
        users::User,
    },
};

const DEFAULT_MAL_BASE_URL: &str = "https://api.myanimelist.net/v2";
const HEADER_NAME: &str = "X-MAL-CLIENT-ID";
const MAL_ANIME_FIELDS: &str = "synopsis,media_type,my_list_status,rating,mean,num_episodes,popularity,alternative_titles,genres";
const MAL_MANGA_FIELDS: &str = "synopsis,media_type,my_list_status,mean,num_chapters,num_volumes,popularity,alternative_titles,genres";
const USER_INFO_URL: &str = "https://api.myanimelist.net/v2/users/@me?fields=id,name,picture";
const MAL_TOKEN_URL: &str = "https://myanimelist.net/v1/oauth2/token";

pub struct MalClient {
    client: reqwest::Client,
    user: User,
}

impl MalClient {
    pub async fn fetch_list(
        &self,
        params: &MediaClientParams,
        is_manga: bool,
    ) -> Result<PaginatedResponse> {
        let username = self.user.username.as_str();

        let (endpoint, fields) = if is_manga {
            ("mangalist", MAL_MANGA_FIELDS)
        } else {
            ("animelist", MAL_ANIME_FIELDS)
        };

        let base_url = format!("{DEFAULT_MAL_BASE_URL}/users/{username}/{endpoint}");

        let norm_status = params
            .status
            .as_deref()
            .and_then(|s| MalStatus::try_from((s, is_manga)).ok());

        let mut req_builder = self.client.get(&base_url).query(&[("fields", fields)]);

        if let Some(status) = norm_status {
            req_builder = req_builder.query(&[("status", status)]);
        }
        if let Some(sort) = &params.sort {
            req_builder = req_builder.query(&[("sort", sort)]);
        }
        req_builder = req_builder.query(&[
            ("limit", params.limit.unwrap_or(50)),
            ("offset", params.offset.unwrap_or(0)),
        ]);

        if let Some(access_token) = self.user.access_token.as_deref() {
            req_builder = req_builder.bearer_auth(access_token);
        }

        req_builder = req_builder.header(HEADER_NAME, env!("MAL_CLIENT_ID"));

        let resp = req_builder.send().await.to_loco_err()?;

        debug!("response received from {:?}", resp.url());

        if !resp.status().is_success() {
            return Err(Error::Message(format!(
                "Error fetching users for MAL with status {}",
                resp.status()
            )));
        }

        let res = resp.json::<MalResponse>().await.to_loco_err()?;

        Ok(res.into())
    }
}
#[async_trait]
impl MediaClient for MalClient {
    fn new(client: &reqwest::Client, user: &User) -> Self {
        Self {
            client: client.clone(),
            user: user.clone(),
        }
    }

    async fn get_anime_list(&self, params: &MediaClientParams) -> Result<PaginatedResponse> {
        Self::fetch_list(self, params, false).await
    }

    async fn get_manga_list(&self, params: &MediaClientParams) -> Result<PaginatedResponse> {
        Self::fetch_list(self, params, true).await
    }

    async fn validate_new_user(&self, access_token: &str) -> Result<User> {
        let resp = self
            .client
            .get(USER_INFO_URL)
            .header(HEADER_NAME, env!("MAL_CLIENT_ID"))
            .bearer_auth(access_token)
            .send()
            .await
            .to_loco_err()?;

        let user: Value = resp.json().await.to_loco_err()?;

        let username = user["name"]
            .as_str()
            .ok_or(Error::Message("Error getting username".into()))?
            .into();

        let new_user = User {
            username,
            provider_id: user["id"].as_i64().map(|id| id.to_string()),
            avatar_url: user["picture"].as_str().map(|url| url.into()),
            provider: MediaProvider::MAL,
            access_token: Some(access_token.into()),
            is_sandbox: false,
            ..Default::default()
        };

        Ok(new_user)
    }

    async fn exchange_oauth_token(&self, code: &str, code_verifier: &str) -> Result<String> {
        let redirect_uri = option_env!("HOSTED_AUTH_PAGE").unwrap_or(DEFAULT_HOSTED_AUTH_PAGE);
        let resp = self
            .client
            .post(MAL_TOKEN_URL)
            .header("Access-Control-Allow-Origin", MAL_TOKEN_URL)
            .form(&json!({
                "client_id": env!("MAL_CLIENT_ID"),
                "code": code,
                "code_verifier": code_verifier,
                "grant_type": "authorization_code",
                "redirect_uri": redirect_uri
            }))
            .send()
            .await
            .to_loco_err()?;

        debug!("exchange_oauth_token: resp={:#?}", resp);

        if resp.status().is_success() {
            let data: Value = resp.json().await.to_loco_err()?;
            return Ok(data["access_token"].as_str().unwrap_or_default().into());
        }

        Err(Error::Message(
            "Error exchanging oauth token for MAL".into(),
        ))
    }
}

File: src/adapters/mal/mal_client.go
package mal

import (
	"encoding/json"
	"fmt"

	"komorebi-server/configs"
	"komorebi-server/src/dto"
	"komorebi-server/src/models"

	"github.com/rs/zerolog/log"
	"github.com/tidwall/gjson"
	"resty.dev/v3"
)

const (
	DEFAULT_MAL_BASE_URL = "https://api.myanimelist.net/v2"
	HEADER_NAME          = "X-MAL-CLIENT-ID"
	MAL_ANIME_FIELDS     = "synopsis,media_type,my_list_status,rating,mean,num_episodes,popularity,alternative_titles,genres"
	MAL_MANGA_FIELDS     = "synopsis,media_type,my_list_status,mean,num_chapters,num_volumes,popularity,alternative_titles,genres"
	USER_INFO_URL        = "https://api.myanimelist.net/v2/users/@me?fields=id,name,picture"
	MAL_TOKEN_URL        = "https://myanimelist.net/v1/oauth2/token"
	CORS_HEADER          = "Access-Control-Allow-Origin"
)

type MalClient struct {
	client *resty.Client
	user   *models.User
}

func NewMalClient(client *resty.Client, user *models.User) *MalClient {
	return &MalClient{client: client, user: user}
}

func (c *MalClient) fetchList(params *dto.MediaClientParams, isManga bool) (dto.PaginatedResponse, error) {
	var endpoint string
	var fields string
	if isManga {
		endpoint = "managalist"
		fields = MAL_MANGA_FIELDS
	} else {
		endpoint = "animelist"
		fields = MAL_ANIME_FIELDS
	}

	url := DEFAULT_MAL_BASE_URL + "/users/" + c.user.Username + "/" + endpoint

	req := c.client.R()

	if params != nil {
		if params.Status != nil {
			if malStatus, err := ParseMalStatus(*params.Status, isManga); err == nil {
				req.SetQueryParam("status", string(malStatus))
			} else {
				req.SetQueryParam("status", *params.Status)
			}
		}
		if params.Sort != nil {
			req.SetQueryParam("sort", *params.Sort)
		}
		if params.Limit == nil {
			req.SetQueryParam("limit", "50")
		} else {
			req.SetQueryParamAny("limit", *params.Limit)
		}
		if params.Offset == nil {
			req.SetQueryParam("offset", "0")
		} else {
			req.SetQueryParamAny("offset", *params.Offset)
		}
	} else {
		req.SetQueryParam("limit", "50")
		req.SetQueryParam("offset", "0")
	}

	if c.user.AccessToken != nil {
		req.SetAuthToken(*c.user.AccessToken)
	}

	req.SetQueryParam("fields", fields)
	req.SetHeader(HEADER_NAME, configs.GetConfig().Env.MalClientId)

	resp, err := req.Get(url)
	if err != nil || resp.StatusCode() != 200 {
		log.Err(err).Str("request", req.CurlCmd()).Bool("isManga", isManga).Msg("Failed to get mal list")
		return dto.PaginatedResponse{}, fmt.Errorf("%w response %s", err, resp.Status())
	}

	var malResp MalResponse
	err = json.Unmarshal(resp.Bytes(), &malResp)
	if err != nil {
		log.Err(err).Str("response", resp.String()).Msg("Failed to parse response in get mal list")
		return dto.PaginatedResponse{}, err
	}
	return malResp.ToPaginatedResponse(), nil
}

func (c *MalClient) GetAnimeList(params dto.MediaClientParams) (dto.PaginatedResponse, error) {
	return c.fetchList(&params, false)
}

func (c *MalClient) GetMangaList(params dto.MediaClientParams) (dto.PaginatedResponse, error) {
	return c.fetchList(&params, true)
}

func (c *MalClient) ValidateNewUser(accessToken string) error {
	req := c.client.R().SetHeader(HEADER_NAME, configs.GetConfig().Env.MalClientId).SetAuthToken(*c.user.AccessToken)
	resp, err := req.Get(USER_INFO_URL)
	if err != nil || resp.StatusCode() != 200 {
		log.Err(err).Any("request", req.Body).Msg("Failed to ValidateNewUser in MAL")
		return fmt.Errorf("%w response %s", err, resp.Status())
	}

	name := gjson.GetBytes(resp.Bytes(), "name")
	id := gjson.GetBytes(resp.Bytes(), "id")
	picture := gjson.GetBytes(resp.Bytes(), "picture")

	if !name.Exists() {
		log.Err(err).Str("response", resp.String()).Msg("Failed to unmarshal resp")
		return err
	}

	c.user.Username = name.String()
	c.user.ProviderId = new(id.String())
	c.user.AvatarUrl = new(picture.String())
	c.user.Provider = dto.MediaProviderMAL
	c.user.AccessToken = &accessToken
	c.user.IsSandbox = false

	return nil
}

func (c *MalClient) ExchangeOauthToken(code, codeVerifier string) (string, error) {
	req := c.client.R().SetHeader(CORS_HEADER, MAL_TOKEN_URL).SetFormData(map[string]string{
		"client_id":     configs.GetConfig().Env.MalClientId,
		"code":          code,
		"code_verifier": codeVerifier,
		"grant_type":    "authorization_code",
		"redirect_uri":  configs.GetConfig().Env.DefaultHostedAuthPage,
	})

	resp, err := req.Post(MAL_TOKEN_URL)
	if err != nil {
		log.Err(err).Str("request", req.CurlCmd()).Msg("Failed to send req for ExchangeOauthToken")
		return "", err
	}

	result := gjson.GetBytes(resp.Bytes(), "access_token")
	if !result.Exists() {
		log.Err(err).Str("response", resp.String()).Msg("Failed to unmarshal resp")
		return "", err
	}

	return result.String(), nil
}
